internal/fs/fs.go

```diff
@@ -264,6 +264,89 @@ func genTestFilename(str string) string {
 	}, str)
 }
 
+var errPathNotDir = errors.New("given path is not a directory")
+
+// ReadActualFilenames is used to determine the actual file names in given directory.
+//
+// On case sensitive file systems like ext4, it will check if those files exist using
+// `os.Stat` and return a map with key and value as filenames which exist in the folder.
+//
+// Otherwise, it reads the contents of the directory
+func ReadActualFilenames(dirPath string, names []string) (map[string]string, error) {
+	actualFilenames := make(map[string]string, len(names))
+	if len(names) <= 0 {
+		// This isn't expected to happen for current usage.
+		// Adding edge case handling, maybe useful in future
+		return actualFilenames, nil
+	}
+	// First, check that the given path is valid and it is a directory
+	dirStat, err := os.Stat(dirPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read actual filenames")
+	}
+
+	if !dirStat.IsDir() {
+		return nil, errPathNotDir
+	}
+
+	// Ideally, we would use `os.Stat` for getting the actual file names
+	// but that returns the name we passed in as an argument and not the actual filename.
+	// So we are forced to list the directory contents and check
+	// against that. Since this check is costly, we do it only if absolutely necessary.
+	caseSensitive, err := IsCaseSensitiveFilesystem(dirPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read actual filenames")
+	}
+	if caseSensitive {
+		// There will be no difference between actual filename and given filename
+		// So just check if those files exist.
+		for _, name := range names {
+			_, err := os.Stat(filepath.Join(dirPath, name))
+			if err == nil {
+				actualFilenames[name] = name
+			} else if !os.IsNotExist(err) {
+				// Some unexpected err, return it.
+				return nil, errors.Wrap(err, "failed to read actual filenames")
+			}
+		}
+		return actualFilenames, nil
+	}
+
+	dir, err := os.Open(dirPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read actual filenames")
+	}
+	defer dir.Close()
+
+	// Pass -1 to read all files in directory
+	files, err := dir.Readdir(-1)
```

Comment:
it should be quite a bit cheaper to use `Readdirnames()` instead of `Readdir()`.

Reply:
I don't think it will be cheaper. I skimmed through the source - https://golang.org/src/os/dir_windows.go. It actually uses `dir.Readdir` under the hood and only returns the names. By using `dir.Readdir`, we can avoid file name comparisons for non-regular(`file.Mode().IsRegular()`) files. Also, 1 less loop 😛

Reply:
that's the windows implementation, but the unix implementation is the reverse - `Readdir` relies on `Readdirnames`. when we do it this way, both platforms are slow; when we do it with `Readdirnames`, only windows is slow.

we could benchmark it, but i imagine that constant factors of retrieving full file stat info for each node will drastically outweigh the reduction in N from skipping a some subset of the names.

Reply:
Oh. Okay. I saw https://golang.org/src/os/dir_plan9.go which was same as windows. Didn't know unix was different. I am guessing the unix one is used on macOS as well. Changing this to use `Readdirnames`.

Reply:
i do appreciate the thoroughness!

Reply:
Ha ha.. That was because all the code I saw was super through. Very very good code comments etc.